src/google/google-calendars/googlecalendarsyncadaptor.cpp

```diff
@@ -1062,13 +1108,59 @@ void GoogleCalendarSyncAdaptor::updateLocalCalendarNotebookEvents(int accountId,
             event->endUpdates();
             m_storageNeedsSave = true;
         } else {
-            // add a new local event
-            remoteAdded++;
-            KCalCore::Event::Ptr event = KCalCore::Event::Ptr(new KCalCore::Event);
-            jsonToKCal(eventData, event, m_icalFormat); // direct conversion
-            m_calendar->addEvent(event, googleNotebook->uid());
+            // add a new local event for the remote addition.
+            KCalCore::Event::Ptr event;
+            if (recurrenceId.isValid()) {
+                // this is a persistent occurrence for an already-existing series.
+                SOCIALD_LOG_DEBUG("Persistent occurrence added remotely:" << eventId);
+                KCalCore::Event::Ptr parentEvent = allMap.value(parentId);
+                if (parentEvent.isNull()) {
+                    // it might have been newly added in this sync cycle.  Look for it from the calendar.
+                    QString parentEventUid = m_recurringEventIdToKCalUid.value(accountId).value(parentId);
+                    parentEvent = parentEventUid.isEmpty() ? parentEvent: m_calendar->event(parentEventUid, KDateTime());
+                    if (parentEvent.isNull()) {
+                        SOCIALD_LOG_ERROR("Cannot find parent event:" << parentId << "for persistent occurrence:" << eventId);
+                        m_syncSucceeded[accountId] = false;
+                        return;
+                    }
+                }
+
+                // dissociate the persistent occurrence, add it, and then update it.
+                event = m_calendar->dissociateSingleOccurrence(parentEvent, recurrenceId, recurrenceId.timeSpec()).staticCast<KCalCore::Event>();
+                if (event.isNull()) {
+                    SOCIALD_LOG_ERROR("Could not dissociate occurrence from recurring event:" << parentId << recurrenceId.toString());
+                    m_syncSucceeded[accountId] = false;
+                    return;
+                }
+
+                // mkcal API is strange, we have to immediately add it without applying modifications, and then apply updates later.
```

Comment:
Curious what goes wrong here, calendarworker in qml plugin sets data and then adds.